### DISCUSSION TITLE
Nested validation + yup default & custom resolvers

### DISCUSSION THREAD

**Discussion by mburisin** [react-hook-form/react-hook-form#8954] [Q&A]
Hi,

first of all many thanks for great validation library, it's really awesome! 

I have been looking into validations for couple of days and used RHF on simple structures, I have just now begun looking into more complex use cases and started to hit some problems. I'm not sure, if it could be issue in RHF or I could do something differently to solve some specific cases.

I have prepared Sandbox here: https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-1-nvg5d6

I used this as reference https://codesandbox.io/s/4k9pt

I'm using Yup in the project, I need to do some deep structures, for example:

![image](https://user-images.githubusercontent.com/38759218/187646018-869f3408-b188-4a93-8910-556d9ad760f3.png)

![image](https://user-images.githubusercontent.com/38759218/187643661-321cd2b5-4b6b-4448-b3c0-07dfe12d6b80.png)

I would like to be able to show all errors in nested structure, as seen in screenshot above.

### Resolvers
I tried to use default RHF Yup resolver https://github.com/react-hook-form/resolvers at first, but I had some problems with it - see next section, so I tried to write custom one to solve my needs, but it brought another set of problems.

In Sandbox in App.tsx https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-1-nvg5d6?file=/src/App.tsx, there is switcher on line 26: 

const useDefaultValidationResolver = true;

if set to false, custom resolver will be used instead of default one.

### Default Yup resolver

**a. Shows only first error in node**

Problem is, that default RHF Yup resolver shows only first error in the node.

Generated error structure if validation failes on metadata.min(2) is:

![image](https://user-images.githubusercontent.com/38759218/187645066-c19157d1-f5c6-4e2e-b7ff-04b77c2a7576.png)

If min rule passes, error structure changes to array:

![image](https://user-images.githubusercontent.com/38759218/187647604-fc1e563a-029b-4695-a110-8fdd4d93989d.png)

There are some edge cases / possible issues - see d. Mixed error stucture (Array / Object). 

**b. Error triggered on change**

Steps to reproduce:

1. click Add Metadata
2. click Remove node

Result:

Error is shown. 

Mode in form is "onSubmit", so I would expect error would not propagated to formState.

When more items are added, the error do not disappear. 

I would expect, this would have to be handled manually via trigger('metadata') and error would not be shown at all if not triggered (trigger or submit).

**c. Trigger validates whole node including children**

Steps to reproduce:

1. click Add Metadata
2. click triggerArray

Result:

All children nodes are validated. It makes sense in some cases, but it would be great to have option to trigger just node element.

For example I wanted to trigger('metadata') on array append, but it caused new array item added to be shown with errors which I don't like in this case.

Is there any way in RHF how to achieve it?

**d. Mixed error stucture (Array / Object)**

Steps to reproduce:

1. submit form -> min error triggered
2. click Add Metadata so there are 2 items in array
3. focus and blur any input in array
4. click getFormState

![image](https://user-images.githubusercontent.com/38759218/187649457-29cf45d4-fe85-40d8-8762-c9e6dd28db5d.png)

Metadata stays as object, but errors in nested structure are inserted as object properties. 

I could probably use manual clearErrors, or trigger to work around this but this uncertain errors structure is not ideal for my use case.

Anyway the main problem for me is a) that Resolver validates only top level, so I tried to write custom resolver, but I have problems with that as well.

### Custom Yup resolver

I wrote custom resolver to modify errors structure to show all errors.

I added at each nested level 'children' object / array to allow for nested errors with fixed structure:

![image](https://user-images.githubusercontent.com/38759218/187653147-3b1fbf17-8f9e-4a74-9d0f-98730ace2397.png)

**1. Custom error structure breaks validate on blur**

Steps to reproduce:

1. In Sandbox in App.tsx https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-1-nvg5d6?file=/src/App.tsx, there is switcher on line 26 const useDefaultValidationResolver = true; Set this to false to use custom resolver.
2. press Submit
3. focus and blur input in array

Result:

Validation on nested field on blur didn't happen.

It seems that custom errors structure breaks some functionality in RHF if it's not in expected format. Trigger / submit validation works with this custom structure though.

If there werent problems described in sections above:

If that error trigger on change didn't happen (b. Error triggered on change) and If I could trigger validation only on the node without children (c. Trigger validates whole node including children), I could probably use this custom resolver with different structure (if it does not break something else important in RHF I haven't tested yet). 

I could probably work around this onBlur validation problem with some extra code, but if I could make it work natively, it would be awesome.

### Summary

1. Is there anything I could do differently to achieve nested validation without problems described above? Am I missing something?
2. Are any issues described above possible bugs worth reporting?

Many thanks for any input / hints!

**Comment by leapful** [comment]
@mburisin May be you can use an extra field like "metadataSize" to check and show error message when array field "metadata" had no valid number of item.

For example:
```
const yupSchema = yup.object<FormState>({
  id: yup.number().required(),
  metadataSize: yup.number().when("metadata", {
    is: (data) => {
      return data?.length < 2;
    },
    then: (schema) =>
      schema.required("metadata field must have at least 2 items")
  }),
  metadata: yup.array().of(
    yup.object({
      key: yup.string().min(2).required(),
      value: yup.string().min(2).required()
    })
  )
});
```

Working demo: https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-1-forked-3mw0yv?file=/src/Form.tsx

**Reply by mburisin** [reply]
@leapful Many thanks for this workaround solution! 👍

I have been doing workaround with custom errors so far, but I had lot of extra code which was harder to maintain than this. 

This seems as better solution, however it is still workaround, some deeper structures would be very hard to solve / maintain like this.

I'm not sure, what is the issue internally in RHF and if some of the issues I found would be worth reporting either as issue / bug or improvement suggestion. What do you think?

**Reply by leapful** [reply]
@mburisin : Actually, the problem is not from react-hook-form but form your validation logic with yup. Yup only performs validation with priority on your field array "metadata" first in this case is "min(2)" then if the parent field array is valid, it will start validate items inside the field array, that's how yup works and also it's a common behavior of most schema validator library.

If you cannot remove "min()" or "max()" of the parent field then the child item inside it cannot be validated until the parent is valid.

**Reply by mburisin** [reply]
@leapful I have been looking into this and Yup resolves all nested fields even if top level is not valid.

I have prepared sandbox [here](https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-2-wigfdq?file=/src/App.tsx).

Just press submit and see console log, that the validation doesn't stop on metadata min(2) condition, but it processess all nested errors:

![image](https://user-images.githubusercontent.com/38759218/191682673-ce9bca88-1b54-4324-ada1-f6cfd6e4dfa2.png)

![image](https://user-images.githubusercontent.com/38759218/191695656-ed209d82-eeee-463b-b878-94447400b5c3.png)

In my custom useYupValidationResolver function, see [here](https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-2-wigfdq?file=/utilsValidation.ts:3156-3180), you can see it uses yup validate function with option abortEarly: false, see docs [here](https://github.com/jquense/yup#schemavalidatevalue-any-options-object-promiseinfertypeschema-validationerror).

So I think, this is not root of the problem. Maybe I'm missing something else...

In default react hook form Yup resolver - see [here](https://github.com/react-hook-form/resolvers/blob/master/yup/src/yup.ts#L10), all errors are also processed , but they are written into "flat" structure and last error on path overwrites previous ones.

I see the problems described in first post as possible issues, I'm not sure what is the cause though, how it works in RHF + custom resolvers etc. Just a quick summary:

1. it seems that custom errors object structure does not work properly with RHF, see section in original post **Custom Yup resolver - 1. Custom error structure breaks validate on blur**.  Validate on blur works only with default errors object structure, which is produced by Default RHF Yup resolver, but it supports only default "flat" structure. See section in original post **Default Yup resolver - a. Shows only first error in node** 

2. It seems RHF depends on default fixed errors object structure for some functionality - for example validate on blur. In some edge cases structure is not "exact/predictable", error could be Object, Array or Object with nested items inserted as props (normally array items) to Object - see **a. Shows only first error in node & d. Mixed error stucture (Array / Object)** which is not ideal. I think it would be best, if error structure could be "independent" and just show errors without ties to any other functionality.

3. Validation on array is triggered on add / remove items from array, but it is not consistent (why / when is error triggered - only on remove, but does not disappear on add etc.) See section in original post **Default Yup resolver - b. Error triggered on change)**

Here is suggestions for improvement for triggering nested structures:

4. trigger('metadata') triggers validation on node and all its children. If I want to trigger validation on adding / removing items from array, I must trigger the node metadata, which triggers its children and shows errors on newly added row with form fields which is not good in this use case. Something like trigger('metadata', { exact: true }) to trigger only this node without triggering its children would be great. Yup supports this with validateAt, validateSyncAt.

I know its possible to work around my problems with solution you suggested with extra validation field, but it's not ideal for more complex structures. It would be better if nested structures could be used without need for workarounds.

**Reply by leapful** [reply]
@mburisin : Thanks for your details summary, it help a lot to get familiar with your use case. I've debugged deeply to find out the problem of using yup resolver.

Since yup resolver aggregates errors based on field value type, in this case it's an array, therefore it will be conflicted with yup errors and overrides the error structure.

The only solution I can think of is using array structure but with extra property for the array itself by using custom method of yup, then the structure will be like:

```
const { formState:  { errors } } = useForm();

// errors.metadata is an array
// errors.metadata[0].key = { message: 'Some error' }, errrors.metadata[0].value = { message: 'Some error' }
// then with the help of custom method of yup, parent error will be at "errors.metadata.error = { message: 'Metadata must have at least 2 items' }".
``` 

https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-2-forked-3ffhk7?file=/src/App.tsx

**Reply by mburisin** [reply]
@leapful Many thanks for your patience! 

This is very interesting solution, it has couple of drawbacks though:
1. errors is array with "custom" prop error and is not very obvious - for example console.log(errors) wouldn't show it, but it's there
2. there will be still problem with dynamic triggering on add / remove metadata item, it will trigger the items inside.
3. requires custom method

From this point of view, better workaround is one you suggested [earlier ](https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-1-forked-3mw0yv?file=/src/Form.tsx)with extra field using "when" so I'm gonna go with that for now.

**What I'm trying to find out is whether using custom resolver, code [here](https://codesandbox.io/s/react-hook-form-repeater-fields-with-yup-validation-2-wigfdq?file=/utilsValidation.ts:3156-3180) with custom errors object structure (see nested 'children' object below) is officially supported or not.**

![image](https://user-images.githubusercontent.com/38759218/191731962-505ee421-9fa3-4c98-8c4e-707dee3b0b86.png)

Is it by design this custom error structure should not be used? Because this custom structure works ok when submit is triggered, so I expected it is supported and should work OK with other RHF functions, but for example field revalidation in onBlur mode does not work, which seems like bug.
